Use 'git init' + 'git fetch' instead of 'git clone', + a small cleanup and colorized output

ulfalizer opened the pull request:
Commit messages say it all:

```
cmd: project: Use 'git init' + 'git fetch' instead of 'git clone'

Use this command sequence instead of 'git clone':

  git init <path>
  In <path>:
    git remote add origin <url>
    git [--depth n] fetch origin [<branch-or-SHA>, if clone-depth is used]
    git update-ref refs/heads/manifest-rev <branch-or-SHA>
    git checkout --detach refs/heads/manifest-rev

This has two advantages:

  - No local branches are created automatically, giving us better
    control over the initial state of the repository

  - The final three commands can be reused for fetching changes from an
    existing repository, giving a single code path for all fetches

Let the initial state of a repository be a detached HEAD at
'manifest-rev'. We could also have the initial state have nothing at all
checked out, but it causes commands like 'git rebase' to fail, and might
look confusing.
```

```
cmd: project: Add a _cloned_projects() helper function

This pattern was cropping up in a bunch of places:

  for project in _projects(args):
      if _cloned(project):
          ...

The new helper replaces it with

  for project in _cloned_projects(args):
      ...

Suggested by Marti Bolivar.
```

```
cmd: project: Colorize informational messages

Print informational messages from West project commands in bright green,
to make it easier to distinguish them from commands output (usually from
Git).

Disable colorized output if stdout is not a terminal, or if we're
running on Windows (ANSI color escapes haven't been tested on Windows
yet, and some googling suggests there might be trouble on Windows 10).

This commit adds a general _print_color() logging helper, so far only
used by log.inf(). _print_color() supports the same arguments as the
standard print() function. A color= argument could be added later for
selecting the color, if it turns out to be useful.
```

Fixes #25

ulfalizer commented on src/west/cmd/project.py:
```diff
@@ -722,7 +724,17 @@ def _inf(project, msg):
-    log.inf('=== ' + _expand_shorthands(project, msg))
+    msg = '=== ' + _expand_shorthands(project, msg)
+
+    if platform.system() != "Windows" and os.isatty(sys.stdout.fileno()):
```
This assumes that `log.inf()` will be going to stdout if stdout is a terminal. Not sure if that's an iffy assumption.

mbolivar replied:
Wouldn't it make more sense to put this in log itself? Maybe enabling color with a kwarg. That way other users can benefit as well.

ulfalizer replied:
Sounds good. Will fix.

ulfalizer replied:
Fixed

mbolivar commented on src/west/cmd/project.py:
```diff
@@ -480,56 +480,67 @@ def _manifest_path(args):
-        _inf(project, msg)
-        _git_base(project, cmd)
+        # Qualify branch names with refs/heads/, just to be safe. Just the
+        # branch name is likely to work as well though.
```
Seems worth testing this with a repository that combines clone-depth with a SHA revision as well

ulfalizer replied:
Would break for most servers I think, because they don't allow fetching an arbitrary SHA (though it works if you already have the commit and it's reachable from some branch/tag you've fetched).

mbolivar replied:
Agreed, thanks for the references and extra digging

ulfalizer commented on src/west/cmd/project.py:
```diff
@@ -480,56 +480,67 @@ def _manifest_path(args):
+        _inf(project, 'Fetching changes for (name-and-path)')
+
+        # If 'clone-depth' is not specified, fetch all branches on the
+        # remote. This gives a more usable repository.
```
Shouldn't break as long as the SHA is reachable from any branch on the remote, which is probably reasonable assumption.

The comment a bit above this code is relevant here too:

```
# Note: Many servers won't allow fetching arbitrary commits by SHA.
# ...
```

mbolivar replied:
> Shouldn't break as long as the SHA is reachable from a branch on the remote, which is probably a reasonable assumption.

I agree this is an edge case, but why not handle it, since it's simple to do so?

> The comment a bit above this code is relevant here too:
> 
> # Note: Many servers won't allow fetching arbitrary commits by SHA.
> # ...

Sounds like a server bug. The git fetch manpage on my system allows it:

```
           The format of a <refspec> parameter is an optional plus +, followed
           by the source <src>, followed by a colon :, followed by the
           destination ref <dst>. The colon can be omitted when <dst> is
           empty. <src> is typically a ref, but it can also be a fully spelled
           hex object name.
```

If that happens, the user can work around their buggy Git server by providing a branch or tag ref, no?

ulfalizer replied:
> > Shouldn't break as long as the SHA is reachable from a branch on the remote, which is probably a reasonable assumption.
> 
> I agree this is an edge case, but why not handle it, since it's simple to do so?

It's also by design. If you aren't specifying a clone depth, it probably means you're going to actively work on the repository, and a repository with all the branches/tags from the remote available is more useful. SHA revisions aren't special-cased.

> If that happens, the user can work around their buggy Git server by providing a branch or tag ref, no?

I think it's disabled on most servers, for security reasons perhaps. Looks like GitHub disables it on all repositories, though I haven't dug up a better source: https://gist.github.com/brunoguerra/3b1a1fb3d4e076bd8fa1d00046ceaed3.

I get that error for the main Zephyr repo and for Kconfiglib:

    $ git init
    $ git remote add origin https://github.com/ulfalizer/Kconfiglib.git
    $ git fetch origin 78682a8e3c4fefffec281f13168cec5c297827bb
    error: Server does not allow request for unadvertised object 78682a8e3c4fefffec281f13168cec5c297827bb

ulfalizer replied:
Same thing happens when fetching from a repository on the local filesystem, so seems like a Git default in general.

ulfalizer replied:
Think I found some relevant stuff by digging through Git sources (with a "TODO: Document this further", because IT...)

https://github.com/git/git/blob/2f743933341f276111103550fbf383a34dfcfd38/Documentation/technical/http-protocol.txt#L331

Looks like almost no one enables those.

mbolivar replied:
> Looks like almost no one enables those.

I guess our corp's git server is a rare exception. OK, I'll fix this in my repository and drop this complaint for west.

ulfalizer commented:
Moved the colorization to the logging module.

I added a general `_print_color()` helper, which is a bit more robust than the old version. It has the same interface as `print()`. Not sure all the functionality will be needed yet.

carlescufi commented on src/west/log.py:
```diff
@@ -64,3 +71,42 @@ def die(*args, exit_code=1):
     sys.exit(exit_code)
+
+
+def _print_color(*args, sep=' ', end='\n', file=sys.stdout, flush=False):
```
I tested it and it looks nice enough on Windows.

ulfalizer replied:
Did you test it with the particular ANSI escapes used here? Doesn't seem to support bright green (`\x1B[92m`) from looking at that page.

https://en.wikipedia.org/wiki/ANSI_escape_code lists that one as "aixterm (not in standard)" too, but some other things seem to use it. Looks a lot nicer than dark green.

ulfalizer replied:
Wonder if it can wrap `stderr` too.

ulfalizer replied:
Ah, yeah, looks like it does. Hadn't read far enough.

carlescufi replied:
It will remap the bright green to whatever:
![image](https://user-images.githubusercontent.com/12450381/45101031-e0058100-b12a-11e8-88de-2bc4fbfe39b7.png)

ulfalizer replied:
Hmm... might get a brighter color with `\x1B[1m` (bold) too, though that'd probably give bold text on *nix.

Should be solvable at least. Maybe we could look into it after this gets in, to avoid too much stuff in a single PR.

carlescufi replied:
OK, I'll create an issue instead

ulfalizer commented:
Got around to the colorama stuff in this PR too. Went for a super-simple hardcoded version in `log.inf()` for now.

@carlescufi Thanks for the tip!

mbolivar commented on src/west/log.py:
```diff
@@ -40,9 +41,23 @@ def dbg(*args, level=VERBOSE_NORMAL):
+    #
+    # colorama automatically strips the ANSI escapes when stdout isn't a
+    # terminal (by wrapping sys.stdout).
+    if colorize: print(colorama.Fore.LIGHTGREEN_EX, end='')
```
Just a nit -- can you put the print() on its own line? flake8 complains about this.

mbolivar commented on src/west/log.py:
```diff
@@ -40,9 +41,23 @@ def dbg(*args, level=VERBOSE_NORMAL):
+    if colorize: print(colorama.Style.RESET_ALL, end='', flush=True)
 
 
 def wrn(*args):
```
Can you please add a `colorize` kwarg to all of the routines, for consistency?

Some ideas for how to handle it:

- `dbg` ignores it
- `wrn` prints orange
- `err` and `die` prints red (and `die` defaults it to True)

ulfalizer replied:
Went for red for both warnings and errors. Is that okay?

Saw a lot of red `WARNING:`s and `ERROR:`s when trying to un-hose my system today, so might've been inspired by that. :P

mbolivar commented:
Thanks for the colorama integration! That's awesome. Just a couple of minor issues on my side.
